RNA Connection component

sergeymitr opened the pull request:
Partially extracted from #18784 and #18182.
Previous step: #19134

#### Changes proposed in this Pull Request:
- RNA Connection package now fully encapsulates the connection flow.
- Connection Manager (`connection-ui`) uses RNA Connection to showcase the connection flow.

#### Jetpack product discussion
p1HpG7-bEn-p2

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
_It's easier to use Jetpack Monorepo to test this PR._
_Make sure to run `yarn build` from the repo root after checking out the branch._

1. Disconnect Jetpack, connect Jetpack using In-Place flow. Confirm that you get connected successfully, and disconnect it again.
2. Go to "Tools -> Connection Manager", click the "Connect" button. After a little wait for the API request, line "Site Registered" should appear, and the In-Place iframe should show up.
3. Click "Approve" in the iframe. The connection form should disappear, and the status line should say "Site and User Connected".
4. Go to Jetpack and disconnect.
5. Set the constant `define( 'JETPACK_SHOULD_NOT_USE_CONNECTION_IFRAME', true )` to test the Calypso flow.
6. Go back to "Tools -> Connection Manager" and click the "Connect" button. After a little wait, you should get redirected to Calypso.
7. Complete the connection flow. You will be redirected to Jetpack Dashboard and not the Connection Manager. That's a known issue and will be addressed later.
8. Disconnect Jetpack again, remove the `JETPACK_SHOULD_NOT_USE_CONNECTION_IFRAME` constant.
9. Go to "Tools -> Connection Manager" and click "Connect". Confirm that the In-Place iframe and the line "Site Registered" appear. Then reload the page.
10. After reloading the page, you should see the "Site Registered" line and the "Connect" button. Click the button and finish the connection flow.

sergeymitr committed:
RNA Connection Component and Connection UI (partially implemented).

sergeymitr committed:
Merge branch 'master' into add/rna-connection

sergeymitr committed:
Registration and user connection in RNA Connection and Connection UI.

sergeymitr committed:
Merge branch 'master' into add/rna-connection

sergeymitr committed:
Connection flow wrap up.

sergeymitr committed:
Merge branch 'master' into add/rna-connection

sergeymitr committed:
Add changelog files

sergeymitr committed:
Update package versions.

sergeymitr committed:
Add Jetpack changelog

sergeymitr committed:
Update connection-ui dependency versions

sergeymitr committed:
Update the Jetpack's composer.lock

sergeymitr committed:
Fix Jetpack GUI test warning

sergeymitr committed:
RNA Connection automated testing.

sergeymitr committed:
Adjust dependency versions

sergeymitr committed:
Minor flow fix and docs.

sergeymitr committed:
Make `connection-ui` require the RNA Connection package.

samiff commented on projects/packages/connection-ui/_inc/components/admin/index.jsx:
```diff
@@ -0,0 +1,80 @@
+/**
+ * The Connection IU Admin App.
+ *
+ * @returns {object} The header component.
```
```suggestion
 * @returns {object} The Admin component.
```

sergeymitr replied:
Good catch, thanks 👍

samiff reviewed (commented):
With these changes pulled down in the monorepo, when I'm attempting to reconnect the site using the normal wp-admin flow, I see some notices:

```
PHP Notice:  Undefined index: shared_salt in /usr/local/src/jetpack-monorepo/projects/plugins/jetpack/json-endpoints/jetpack/class.jetpack-json-api-sync-endpoint.php on line 116
```

--- 

When trying to visit Tools > Connection Manager I was getting a `Module build failed (from ./node_modules/@automattic/mini-css-extract-plugin-with-rtl/dist/loader.js):` but nuking my node_modules/vendor/build folders and building `connection-ui` again fixed it.

Was able to successfully connect site & user there 🥳 

---

> Complete the connection flow. You will be redirected to Jetpack Dashboard and not the Connection Manager. That's a known issue and will be addressed later.

When testing the `JETPACK_SHOULD_NOT_USE_CONNECTION_IFRAME`, that worked to redirect me to Calypso for the connection approval, but then took me to the Calypso Plans page (my test site does have a paid plan), rather than directing me back to wp-admin. The connection did work ok though. Would a8c proxy have any effect here?

---

Other steps tested well for me @sergeymitr  👍 

A few other thoughts:

1. Will there be a readme for the `connection-ui` package?
2. Would you reinforce my understanding of the different connection packages:
    * `projects/packages/connection` - lower level bits required for a Jetpack connection.
    * `projects/js-packages/connection` - higher level React components used to build connection flows.
    * `projects/packages/connection-ui` - higher level shared UI for _all the things_ wanting to use a Jetpack connection utilizing the `js-packages/connection` components.
3. If my understanding of the `connection-ui` package is correct (a shared UI for anything that wants a JP connection), maybe you could help me understand a bit of implementation detail:
    * Given an example of a standalone plugin like the [Backup](https://github.com/Automattic/jetpack/pull/19332) plugin, ideally how would the plugin know the connection state, prompt a user to connect, etc.?

sergeymitr committed:
Fixing typo in the `Admin` component

Co-authored-by: Samiff <samiff@users.noreply.github.com>

samiff committed:
Tighten no-unused-var ignores

sergeymitr committed:
Fixinig the Calypso flow:
- skip the "Plans" page
- return users back to Connection Manager

sergeymitr committed:
Readme file for the `connection-ui` package.

sergeymitr commented:
Hi @samiff, thanks for taking a look!

> PHP Notice:  Undefined index: shared_salt

I wasn't able to reproduce the issue, can you please double check? This appears to be a sync issue, and shouldn't be caused by this PR.

> but then took me to the Calypso Plans page (my test site does have a paid plan), rather than directing me back to wp-admin.

I adjusted Calypso to skip the "Plans" page entirely, thanks for pointing this out.

> Will there be a readme for the connection-ui package?

Added 👍 

> Would you reinforce my understanding of the different connection packages: 

You understanding of `packages/connection` and `js-packages/connection` is correct.
`packages/connection-ui` has started as a single page shared between plugins to manage the connection.
Right now that concept is a bit outdated, and this package will eventually implement the "My Jetpack" section.

So at the moment `packages/connection-ui` serves as an example of using the RNA Connection (`js-packages/connection`). Because RNA Connection is not used anywhere else, and we need a way to test it.

> Given an example of a standalone plugin like the Backup plugin, ideally how would the plugin know the connection state, prompt a user to connect, etc.?

The Backup plugin will implement its own React state and initialize the RNA Connection similarly to this: https://github.com/Automattic/jetpack/blob/0580caaae65f0f45fbd51216d2d95b03b9082483/projects/packages/connection-ui/_inc/components/admin/index.jsx

samiff commented:
> This appears to be a sync issue, and shouldn't be caused by this PR.

Good to know, after cleaning everything and rebuilding with this branch I could not reproduce those warnings.

> I adjusted Calypso to skip the "Plans" page entirely

That is working for me now, redirected to the Connection Manager screen after making the user connection.

> The Backup plugin will implement its own React state and initialize the RNA Connection

Thanks for adding the readme and explaining this a bit more, I think I was envisioning a more "hands-off" approach for consumers of a JP connection, but with the example you referenced I see how that will provide for more flexibility 👍

sergeymitr commented:
> I think I was envisioning a more "hands-off" approach for consumers of a JP connection

The problem is that we have two separate React states:
- the plugin (or `connection-ui` package) state
- the RNA Connection (`js-packages/connection`) state

We can't really merge them, so the only way for us to pull the data into the JS package is via properties, that's why there are quite a few of them.
But other than that, the RNA Connection JS package manages its own state, pulls data from the API, etc.

As we work with the Backup plugin, we might be able to simplify it a bit more. We'll see where this road leads us 🙂

sergeymitr committed:
Merge branch 'master' into add/rna-connection

sergeymitr committed:
Fix the Calypso flow redirect.

sergeymitr committed:
Version bump

sergeymitr committed:
Merge branch 'master' into add/rna-connection

sergeymitr committed:
Backup build fixes

jeherve commented on projects/packages/connection-ui/_inc/components/admin/index.jsx:
```diff
@@ -0,0 +1,82 @@
+/**
+ * External dependencies
+ */
+import React, { useCallback } from 'react';
```
Would it be possible to use `@wordpress/element` as much as possible in the package, instead of using React directly? I'm sorry if this was discussed already in the past.

sergeymitr replied:
We discussed it very briefly, and it appeared that there isn't much value to us in adding another abstraction level.
I think it would only further complicated the code with no actual need. We could add them later if we need them.
That's open for discussion though.

sergeymitr replied:
Created a task 1200295095943283-as-1200295095943289.

jeherve commented on projects/packages/connection-ui/_inc/components/admin/style.scss:
```diff
@@ -0,0 +1,18 @@
+#jetpack-connection-ui-container {
+  max-width: 800px;
```
I'd suggest using something like Prettier to make sure that all your spaces are converted into tabs on save, so we remain consistent with the rest of the codebase.

sergeymitr replied:
Fixed now, thanks!

jeherve commented on projects/packages/connection/src/class-rest-connector.php:
```diff
@@ -129,6 +129,15 @@ public function __construct( Manager $connection ) {
 						'type'        => 'string',
+						'required'    => true,
+					),
+					'no_iframe'          => array(
+						'description' => 'Disable In-Place connection flow and go straight to Calypso',
+						'type'        => 'boolean',
+					),
+					'redirect_uri'       => array(
+						'description' => 'URI of the admin page where the user should be redirected after connection flow',
```
Do we want those things to be translatable, here and above?

sergeymitr replied:
Good catch, fixed!

kbrown9 commented on projects/packages/connection-ui/src/class-initial-state.php:
```diff
@@ -35,8 +37,21 @@ public function __construct() {
+				'registrationNonce' => wp_create_nonce( 'jetpack-registration-nonce' ),
+			),
+			'connectionData'   => array(
+				'doNotUseConnectionIframe' => $is_safari || User_Agent_Info::is_opera_desktop() || Constants::is_true( 'JETPACK_SHOULD_NOT_USE_CONNECTION_IFRAME' ),
```
Should we check the `jetpack_use_iframe_authorization_flow` filter here as well? That filter is described [here](https://github.com/Automattic/jetpack/blob/2cdf4dc27aa4cd98207693d798ff8968fc8876ad/projects/packages/connection/src/class-manager.php#L810).

sergeymitr replied:
Added, thanks!

sergeymitr committed:
REST registration endpoint description translation.

sergeymitr committed:
Utilize the `jetpack_use_iframe_authorization_flow` filter for Connection UI.

sergeymitr committed:
Merge branch 'master' into add/rna-connection

sergeymitr committed:
Code styling.

sdixon194 commented:
Did a lot of testing and this works very well for me in all test cases. Also tried a mix of site connection and user authentication split between the connection manager and the jetpack dashboard and it all seemed to work well.

I know there's an ongoing discussion about the particular size of the components ( p1620096252118700/1620074116.112700-slack-CBG1CP4EN ). @sergeymitr do you expect to make any of those changes before merging, or is that something that will be iterated upon later?

sergeymitr commented:
Hi @sdixon194, thanks for taking a look!
Other improvements will probably go into the next PR, and I'd like to merge this one as is unless somebody's considering this a blocker, or something else unfolds.

sdixon194 commented:
🚢 !

kraftbj commented on package.json:
```diff
@@ -16,8 +16,8 @@
 		"build-jetpack": "cd projects/plugins/jetpack && yarn build",
-		"build-jetpack-backup": "cd projects/plugins/jetpack-backup && yarn build",
-		"build-jetpack-backup-concurrently": "cd projects/plugins/jetpack-backup && yarn build-concurrently",
+		"build-jetpack-backup": "cd projects/plugins/backup && yarn build",
```
100% absolutely not a blocker (let's merge this!),  but is there a need for this vs using `jetpack build plugins/backup` command?

Ideally, we can use the CLI and not require adding these steps for every project.

sergeymitr replied:
I agree, it might be excessive, although I'm a bit behind on understanding of the modern Jetpack devops 🙂 
When I recommended @samiff to use it [in another PR](https://github.com/Automattic/jetpack/pull/19332#issuecomment-821680665), I took the `build-jetpack` commands as an example.
If they're no longer needed, we could remove them in #19716, which will be merged after this one.

sergeymitr committed:
Merge branch 'master' into add/rna-connection

jeherve commented on projects/js-packages/connection/package.json:
```diff
@@ -1,16 +1,18 @@
 	"license": "GPL-2.0-or-later",
 	"dependencies": {
 		"@wordpress/browserslist-config": "3.0.1",
+		"@wordpress/components": "9.2.6",
```
This seems to be an old version of the component. Latest available is `13.0.3`. Is it an option to use the latest available?

mmtr replied:
This seems to be causing some CI failures like https://github.com/Automattic/jetpack/pull/19745/checks?check_run_id=2516520038:

```
[test-adminpage] Error: Cannot find module '@wordpress/components'
```

I guess the `@wordpress/components` dep should be also declared in the root `package.json`? That's what we did in https://github.com/Automattic/jetpack/pull/19134 for the `@wordpress/i18n` dep.

mmtr replied:
Opened https://github.com/Automattic/jetpack/pull/19758 to see if the CI failures get solved.

sergeymitr replied:
Thanks @mmtr, it looks like your PR solved the issue!
There's a linter failing though, you probably need to run `yarn install` in the project root, then commit updated `yarn.lock`.

sergeymitr replied:
Hi @jeherve, that's a valid point.
I had to downgrade it at some point to make it match the Jetpack version, otherwise yarn would run into a bunch of non-explainable problems.
We will likely be able to update it separately from Jetpack after switching from Yarn to Node.
Or we could update `@wordpress/components` in both Jetpack and RNA Connection.

sergeymitr replied:
Added a task to update `@wordpress/components`: 1200295095943283-as-1200307886010305
